File: TwitterPluginSource.m
//
//  TwitterPluginSource.m
//  TwitterPlugin
//
//  Created by mootoh on 2008.09.13
//  License: revised BSD
//
#import "TwitterPluginSource.h"
#import "TwitterPluginShared.h"

static int count = 0;

@implementation TwitterPluginSource

- (id) init
{
  if (self = [super init]) {
    // collect friends
    friends_ = [NSMutableArray array];
  }
  return self;
}

- (BOOL) indexIsValidFromDate:(NSDate *)date forEntry:(NSDictionary *)entry
{
  return YES;
}

- (NSImage *) iconForEntry:(NSDictionary *)entry
{
  return [QSResourceManager imageNamed:@"girl_square"];
}

- (NSArray *)objectsForEntry:(NSDictionary *)entry
{
  NSDictionary *dict = [[NSUserDefaultsController sharedUserDefaultsController] values];
  NSString *screen_name = [dict valueForKey:@"TwitterPreference.screenName"];

  NSMutableArray *objects = [NSMutableArray array];

  NSString *base = [NSString stringWithFormat:@"http://twitter.com/statuses/friends/%@.xml?lite=true&", screen_name];

  for (int i=1; true; i++)
  {
    NSMutableArray *friends = [NSMutableArray array];

    NSURL *friends_url = [NSURL URLWithString:[NSString stringWithFormat:@"%@page=%d", base, i]];
    NSError *err = nil;

    NSXMLDocument *doc = [[[NSXMLDocument alloc] initWithContentsOfURL:friends_url
      options:NSXMLDocumentTidyXML error:&err] autorelease];
    if (err) {
      NSLog(@"XML cannot be retrieved and parsed correctly.");
      return nil;
    }

    NSArray *users = [doc nodesForXPath:@"/users/user" error:&err];
    if (err) {
      NSLog(@"XPath failed.");
      return nil;
    }

    if (0 == [users count]) break;

    NSXMLElement *user;
    for (user in users) {
      NSXMLElement *child;
      NSString *screen_name;
      NSString *image_url;

      int count = 0;
      for (child in [user children]) {
        if (count == 2) break;
        NSString *cn = [child name];
        if ([cn isEqualToString:@"screen_name"]) {
          screen_name = [child stringValue];
          count++;
        } else if ([cn isEqualToString:@"profile_image_url"]) {
          image_url = [child stringValue];
          count++;
        }
      }

      NSAssert(screen_name, @"screen_name should exist");
      NSAssert(image_url, @"image_url should exist");
      QSObject *obj = [QSObject objectWithName:screen_name];
      NSAssert(obj, @"QSObject should not be nil.");

      [obj setObject:@"" forType:@"TwitterPluginType"];
      [obj setPrimaryType:@"TwitterPluginType"];
      [obj setObject:image_url forType:@"image_url"];
      [objects addObject:obj];
    }
  }

  // add public one
  QSObject *obj = [QSObject objectWithName:PUBLIC];
  [obj setObject:@"" forType:@"TwitterPluginType"];
  [obj setPrimaryType:@"TwitterPluginType"];
  [objects addObject:obj];

	return objects;
}

- (void) setQuickIconForObject:(QSObject *)object 
{
  [object setIcon:nil];
}

- (BOOL) loadIconForObject:(QSObject *)object {
  if ([[object name] isEqualToString:PUBLIC]) {
    [object setIcon:[QSResourceManager imageNamed:@"girl_square"]];
    return YES;
  }
  NSImage *img = [[[NSImage alloc] initWithContentsOfURL:[NSURL URLWithString:[object objectForType:@"image_url"]]] autorelease];
  if (! img) return false;
  [object setIcon:img];

  return true;
}

@end
// vim:set ft=objc:

File: TwitterPluginSource.rb
require 'osx/cocoa'
require 'open-uri'
require 'rubygems'
require 'json'

class TwitterPluginSource < OSX::QSObjectSource
  @@count = 0
  FRIENDS = File.dirname(__FILE__) + '/friends.dat'

  def initialize
    super
    @friends = File.exist?(FRIENDS) ? Marshal.load(open(FRIENDS)) : []
  end

  def indexIsValidFromDate_forEntry(index, entry)
    true
  end

  def iconForEntry(dict) # dict.keys => name, source, ID, bundle
    OSX::QSResourceManager.imageNamed('girl_square')
  end

=begin
  - (NSString *)identifierForObject:(id <QSObject>)object{
        return nil;
    }
=end

  def name_pass
    screen_name.to_s + ':' + password
  end

  def objectsForEntry(entry)
    dict = OSX::NSUserDefaultsController.sharedUserDefaultsController.values;
    screen_name = dict.valueForKey("TwitterPreference.screenName")
    password    = dict.valueForKey("TwitterPreference.password")

    @friends = []
    objects = []
    count = 1

    while true
      url = 'http://' + screen_name + ':' + password +
            '@twitter.com/statuses/friends/' + screen_name + '.json'

      if count > 1
        url += '?page=' + count.to_s
      end
      count += 1

      friends = JSON.parse(open(url).read)
      friends.each do |friend|
        Shared.logger.info(friend['screen_name'])
        obj = OSX::QSObject.objectWithName(friend['screen_name'])
        obj.setObject_forType('', 'TwitterPluginType')
        obj.setPrimaryType('TwitterPluginType')
        objects.push(obj)
      end

      @friends += friends

      break if (friends.size < 100)
    end
    @friends.uniq!
    Marshal.dump(@friends, open(FRIENDS, 'w'))
    objects
  end

  # Object Handler Methods

  def setQuickIconForObject(object)
    object.setIcon(nil) 
  end

  def loadIconForObject(object)
    return false unless @friends
    return false if @friends.empty?

    Shared.logger.info('1')

    her = @friends.find {|x| x['screen_name'] == object.name.to_s}
    img = OSX::NSImage.alloc.initWithContentsOfURL(OSX::NSURL.URLWithString(her['profile_image_url']))
    return false unless img

    object.setIcon(img)
    true
  end
end
